src/main/scala/temple/DSL/semantics/Validator.scala

```diff
@@ -1,44 +1,81 @@
 package temple.DSL.semantics
 
+import temple.DSL.semantics.NameClashes._
+import temple.DSL.semantics.Validator._
 import temple.ast.AttributeType._
 import temple.ast.{Metadata, _}
+import temple.builder.project.ProjectConfig
 
 import scala.collection.mutable
 import scala.reflect.{ClassTag, classTag}
 
-private class Validator private (templefile: Templefile) {
-  var errors: mutable.Set[String] = mutable.SortedSet()
+private class Validator(templefile: Templefile) {
+  var errors: mutable.Buffer[String] = mutable.Buffer()
 
-  private lazy val allStructs: Iterable[String] = templefile.services.flatMap {
-    case (name, service) => service.structs.keys
+  private val allStructs: Iterable[String] = templefile.services.flatMap {
+    case _ -> service => service.structs.keys
   }
 
-  private lazy val allServices: Set[String] = templefile.services.keys.toSet
+  private val allServices: Set[String] = templefile.services.keys.toSet
 
-  private lazy val allStructsAndServices: Set[String] = allServices ++ allStructs
+  private val allStructsAndServices: Set[String] = allServices ++ allStructs
 
-  private def validateAttributes(attributes: Map[String, Attribute])(context: SemanticContext): Unit = {
-    attributes.keys.foreach { attributeName =>
-      if (attributeName.headOption.forall(!_.isLower))
-        errors += context.errorMessage(s"Invalid attribute name $attributeName, it must start with a lowercase letter,")
+  private var newServices: Map[String, ServiceBlock] = templefile.services
+
+  private val globalRenaming      = mutable.Map[String, String]()
+  def allGlobalNames: Set[String] = globalRenaming.valuesIterator.toSet ++ allStructsAndServices
+
+  private def validateAttributes(
+    attributes: Map[String, Attribute],
+    context: SemanticContext,
+  ): Map[String, Attribute] = {
+    foreachValueWithContext(attributes, context, validateAttribute)
+
+    // This looks like a map, but it isn’t because the accumulator `newAttributes`’ names are accessed during iteration
```

Comment:
Rather than saying what this should or shouldn't look like, maybe explain the map you're building up?